Detect module dependencies when modules are not default-loaded

gcoxmoz opened the pull request:
## Additional Context
If you accept `default_mods => true` you get a -lot- of apache modules automatically.  In my experience it's too many, which has led to us doing `default_mods => []` to reduce the sprawl, and then adding back in what we need.  That leads to a bit of whack-a-mole upon new server buildout, as a lot of this module's code assumes that you have included the default modules and it doesn't need to check.  It also leads to "I'm not sure if I can remove this module" later, if you remove directives.

## Summary
This covers off the worst offenders I've hit in our environment: `mod_dir`, `mod_autoindex`, `mod_expires`, `mod_dav`, and `mod_cgi`.  If you use vhost directory parameters that will invoke these module directives, vhost will try to make sure the module is loaded.  If you're on `default_mods => true`, this is a 'wasted' extra load; if you don't have the module loaded (say, `default_mods => false`) then this brings the module in for you.

## Related Issues (if any)
N/A

## Checklist
- [x] 🟢 Spec tests.
- [ ] 🟢 Acceptance tests.
- [x] Manually verified. (For example `puppet apply`)

gcoxmoz committed:
Ensure mod::dir is added when needed

gcoxmoz committed:
Ensure mod::expires is added when needed

gcoxmoz committed:
Ensure mod::dav is added when needed

gcoxmoz committed:
Ensure mod::autoindex is added when needed

gcoxmoz committed:
Ensure mod::cgi is added when needed

binford2k commented:
Should this magically load dependencies, or should it display errors so people know to correct their code? I'm not sure I have a strong opinion there. Perhaps it could load, but also print an `INFO` message.

gcoxmoz commented:
There's plenty of prior art of the puppet-module "acting on your behalf" in grabbing needed apache-modules.  One for a quick example:  https://github.com/puppetlabs/puppetlabs-apache/blob/1c720758bf4ac605c58a1cb0e1f181566af377b8/manifests/vhost.pp#L2261-L2263 is "if your vhost directive asked for `virtual_docroot` (which defaults to `false`), jump in and grab the apache-module that will make it work."  This parallels what I'm looking for: 'demand an apache module when it is needed'.  The difference is that a lot of the puppet module is built on the assumption that some of the apache modules are always there... and they don't have to be.  Which is actually a bit odd, since the spec test https://github.com/puppetlabs/puppetlabs-apache/blob/1c720758bf4ac605c58a1cb0e1f181566af377b8/spec/defines/vhost_spec.rb#L11 assumes `default_mods => false` while most of the code is built with an unspoken assumption of `default_mods => true`

That all happens without an `INFO` or an explicit apache-module declaration, so IMO autoloading the apache-modules (quietly, as needed) is the way to go here.

ekohl commented on manifests/vhost.pp:
```diff
@@ -2335,10 +2336,49 @@
+
+      if 'dav' in $directory {
+        include apache::mod::dav
+        if $directory['dav'] == 'On' {
```
I'm not sure we really enforce this to be `On`. Could be `on` or a boolean too, right? Would it be better to reverse the logic and first check for `svn` and then:
```suggestion
        if bool2httpd($directory['dav']) == 'On' {
```

ekohl reviewed (commented):
I've been pushing in this direction. Like https://github.com/puppetlabs/puppetlabs-apache/pull/2255 and https://github.com/puppetlabs/puppetlabs-apache/pull/2337. Exactly for the reason you mention: I want to run the server with `default_mods => false` but not think about which modules to include.

My ideal is that we actually define a proper data type for `$directories` and enforce it.

gcoxmoz committed:
Fix dav logic based on suggestions

ekohl reviewed (commented):
I want to let you know that I've essentially stopped reviewing anything that is Perforce maintained since they are very unclear on outside maintainers. There used to be https://puppet.com/ecosystem/trusted-contributors/ but that's now gone.

gcoxmoz committed:
More updates based on review comments

Bikeshedding over one if statement, what are we even doing, no wonder we can't get commits merged.

jst-cyr commented:
@gcoxmoz : Thanks for the PR! Sorry for the long delays in getting to this. I'm going through old PRs and wanted to check if this was still an issue for you? It seems like a helpful change, but wanted to double-check.

gcoxmoz commented:
Yes, still valid.   I re-rebased up to main.

jst-cyr commented:
Running CI for this PR here: https://github.com/puppetlabs/puppetlabs-apache/pull/2641

jst-cyr commented:
Acceptance test results are here: https://github.com/puppetlabs/puppetlabs-apache/actions/runs/31515279283/job/93930997451?pr=2641

Mostly green, with these platforms failing acceptance tests:
- Debian 13: Failed with same `apache2 fails to start via systemd` error as the nightly CI failures against main
- SLES 12: Provisioning succeeded on second run, but still failed with the same errors as seen in nightly CI against main
- SLES 15: Failed with the same `Could not find command '/usr/sbin/apache2ctl` error as the nightly CI failures against main

The PR does not seem to be introducing any new failures from an acceptance testing perspective.

gcoxmoz commented:
So, like, are we just stuck here til someone greens up main?

jst-cyr commented:
@gcoxmoz : The step I ran was to get acceptance tests going, which is part of the review process. From my standpoint, tests for this PR don't show any additional issues being introduced, so I think our acceptance testing gate is probably fine for now. We'll likely need to run these again if more changes happen in main and we need to rebase.

The next step is getting the engineering team to do a review of the changes. That could be some time.
